File: brandon-heisler/day-08/who-likes-it.js
function likes(names) {
  if (names.length === 0) {
    return "no one likes this";
  }

  else if (names.length === 1) {
    return `${names[0]} likes this`;
  }

  else if (names.length === 2) {
    return `${names[0]} and ${names[1]} like this`;
  }

  else if (names.length === 3) {
    return `${names[0]}, ${names[1]} and ${names[2]} like this`;
  }

  else if (names.length >= 4) {
    return `${names[0]}, ${names[1]} and ${names.length - 2} others like this`;
  }

}

/*
were given an array of names. we need to make a function that adds
the names in the given array to a string of text.
we should have a couple strings.
1. if no name, no one likes this
2. if 1 name (name) likes this
3. if 2 or 3 (name, name, name) likes this.
4. if 4+ (first two names) and (#'s') others like this.

we want to return a str.

i can use string literal template to fill in the blanks

*/
/*
  Ember's Feedback:
  - Great work!
  - Excellent pseudo code and execution
*/
// Alternative Solution
function likes(names) {
  switch (names.length) {
    case 0:
      return 'no one likes this';
    case 1:
      return `${names[0]} likes this`;
    case 2:
      return `${names[0]} and ${names[1]} like this`;
    case 3:
      return `${names[0]}, ${names[1]} and ${names[2]} like this`;
    default:
      return `${names[0]}, ${names[1]} and ${names.length - 2} others like this`;
  }
}

File: brandon-heisler/day-38/who_likes_it.py
def likes(names):
    if len(names) == 0:
        return "no one likes this"
    elif len(names) == 1:
        return "{} likes this".format(names[0])
    elif len(names) == 2:
        return "{} and {} like this".format(names[0], names[1])
    elif len(names) == 3:
        return "{}, {} and {} like this".format(names[0], names[1], names[2])
    elif len(names) > 3:
        others_count = len(names) - 2
        return "{}, {} and {} others like this".format(names[0], names[1], others_count)


# im going to make an if statement to check for the length of the given array and format
# elif statements for different conditions and format the string accordingly, attaching
# the names and counts for each case

"""
    Ember's Feedback:
    - Great use of .format! :D
"""

# Alternative Solutions
def likes(names):
    match len(names):
        case 0:
            return "no one likes this"
        case 1:
            return f"{names[0]} likes this"
        case 2:
            return f"{names[0]} and {names[1]} like this"
        case 3:
            return f"{names[0]}, {names[1]} and {names[2]} like this"
        case _:
            return f"{names[0]}, {names[1]} and {len(names) - 2} others like this"


# OR
def likes(names):
    n = len(names)
    return {
        0: "no one likes this",
        1: "{} likes this",
        2: "{} and {} like this",
        3: "{}, {} and {} like this",
        4: "{}, {} and {others} others like this",
    }[min(4, n)].format(*names[:3], others=n - 2)


# OR
def likes(names):
    match names:
        case []:
            return "no one likes this"
        case [a]:
            return f"{a} likes this"
        case [a, b]:
            return f"{a} and {b} like this"
        case [a, b, c]:
            return f"{a}, {b} and {c} like this"
        case [a, b, *rest]:
            return f"{a}, {b} and {len(rest)} others like this"